### PR TITLE
[runtime env] add and remove uri reference in worker pool

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
Currently, the logic of uri reference in raylet is:
- For job level, add uri reference when job started and remove uri reference when job finished.
- For actor level, add and remove uri reference for detached actor only.

In this PR, the logic is optimized to:
- For job level, check if runtime env should be installed eagerly first. If true, add or remove uri reference. 
- For actor level
    * First, add uri reference for starting worker process to avoid that runtime env is gcd before worker registered.
    * Second, add uri reference for echo worker thread of worker process. We will remove reference when worker disconnected.

- Besides, we move the instance of `RuntimeEnvManager` from `node_manager` to `worker_pool`.
- Enable the test `test_actor_level_gc` and add some tests in python and worker pool test.

## Related issue number

https://github.com/ray-project/ray/issues/19602

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_runtime_env_working_dir_2.py
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_working_dir_2.py:320]: Thanks! Can we also enable `test_runtime_env_conda.py::test_actor_level_gc`?

**Review by edoakes — COMMENTED** [review]
Add unit tests to worker_pool_test.cc please

File: src/ray/raylet/worker_pool.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool.cc:36]: why is this needed? also, won't this break the logic in the agent because now the URIs have this unknown prefix?

File: src/ray/raylet/worker_pool.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool.cc:599]: why don't we need the same check in HandleJobStarted?

**Comment by architkulkarni** [comment]
There's a second `runtime_env_manager` in the GCS server.  That one is unchanged by this PR (so it still only counts URIs per-job and per detached actor).  I think this is okay for now, because the only URIs the GCS server cares about are for working_dir and py_modules, which are only specified at `ray.init()` time.  (The GCS cleans up the working_dir and py_modules files that were uploaded from the driver and stored in the internal KV.). But it might be confusing to have two different strategies for refcounting--let me think about if there's a way to improve this in a future PR.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:474]: Is it guaranteed that this only happens after AddURIReference in OnWorkerStarted?  I just want to make sure there isn't a gap between when the reference is removed here and when it's added in OnWorkerStarted.  If there's a gap, then the refcount might become 0 in the gap and we might start deleting the resources

File: python/ray/tests/test_runtime_env_working_dir_2.py
Comment (architkulkarni) [reply python/ray/tests/test_runtime_env_working_dir_2.py:320]: And in the conda test, do you think you could add a test to make sure the GC works when we specify conda in a task?  (It should work, right? because the references will be counted per worker.). Specifically, I'm interested in the following situation:
1. Start a task with conda env A
2. Start an actor with conda env A
3. Kill the actor (and perhaps wait for some time)
4. Start another task with conda env A

With the naive per-actor ref counting, the refcount would go to 0 at step 3, so the conda env would be deleted and step 4 would fail with ModuleNotFoundError (because the second task would run on the same worker as the first task, but the conda environment is gone.)

But with this PR, it should work properly because the worker will contribute to the refcount.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:36]: This (prefix + startup_token) isn't used in the URI name, it's just used as an internal key by the runtime_env_manager.  For example, in other usage, it could be a job_id, or an actor_id.  The key points to a set of URIs so that the runtime_env_mangaer knows which URIs to decrement the refcounts for when the job, actor, or worker exits.

Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:36]: The comment on this line makes sense to me, but I'm not sure what the `startup_token` is actually.  Is it some integer that's guaranteed to be unique for each worker?

File: src/ray/common/runtime_env_manager.cc
Comment (rkooo567) [inline src/ray/common/runtime_env_manager.cc:38]: Can we remove this log? This will make log really spammy when debug log is on.

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:433]: Can we put `state.starting_worker_processes.emplace(` logic and `runtime_env_manager_.AddURIReference(` in the same private method? Seems like they should always be coupled right? Same for `.erase` and `RemoveURIReference `

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:434]: Can we also append the task or actor id? That'll help us debugging issues

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:599]: Can you comment why we have this eager_install check in the if statement? Hard to understand the reasoning without that if you are not familiar with runtime env

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:601]: Is it possible to move `runtime_env != "{}" && runtime_env != ""` to a separate helper function?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Can you add a DebugString to the `runtime_env_manager_ ` and print the ref status to worker_pool's debug string?

Also, it is very hard to reason about if the ref addition / removal is correct by reading code (it is easy for the job, but not for workers). Is it possible to write a docstring somewhere to see where the ref is added and how they are decremented in the big picture? Like the worker lifecycle diagram & when ref is incremented or decremented.

File: python/ray/tests/test_runtime_env_working_dir_2.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_working_dir_2.py:320]: Sounds good. But I have a concern that we have a worker processes cache for normal tasks. If the worker pool don't kill idle workers immediately, the GC will not happened. But I think the task could always be finished. Let me confirm the logic.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:36]: Yep. And I have a concern that more and more modules will use the uri reference in future. So I added a prefix. Actually, it should also work if I remove this prefix.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:599]: image
We already have this check?

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:474]: We can guarantee that this only happens after  AddURIReference  in `StartWorkerProcess`, not `OnWorkerStarted `. If some worker thread register timeout after process started, we will run this RemoveURIReference.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:433]: A little weird because this private method only used by `StartWorkerProcess`. But the erase logic is used in two places now. Make sense overall.

File: src/ray/common/runtime_env_manager.cc
Comment (SongGuyang) [reply src/ray/common/runtime_env_manager.cc:38]: This log is useful for debug now. And The log count is equal to starting worker number probably. Maybe it is acceptable for debug level logs？

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:434]: We can't map one worker process to a task because the worker process can be cached and reused for normal task. Maybe the only thing we can do is appending actor id for dedicated workers? We can guarantee that a dedicated worker occupy a worker process now.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:599]: Yep, I can add a comment. We just add uri reference when runtime env install happens really.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:601]: Make sense.

**Comment by SongGuyang** [comment]
> Can you add a DebugString to the `runtime_env_manager_ ` and print the ref status to worker_pool's debug string?
> 
> Also, it is very hard to reason about if the ref addition / removal is correct by reading code (it is easy for the job, but not for workers). Is it possible to write a docstring somewhere to see where the ref is added and how they are decremented in the big picture? Like the worker lifecycle diagram & when ref is incremented or decremented.

Good suggestion. But can we add this in a independent PR? We have some other changes which blocked by current PR.

**Comment by rkooo567** [comment]
> Good suggestion. But can we add this in an independent PR? We have some other changes which blocked by current PR.

I think we should focus on making correctly working code rather than being hurry for other blockers. And I'd like to have the confidence as a reviewer. 

What about only doing 

> Also, it is very hard to reason about if the ref addition / removal is correct by reading code (it is easy for the job, but not for workers). Is it possible to write a docstring somewhere to see where the ref is added and how they are decremented in the big picture? Like the worker lifecycle diagram & when ref is incremented or decremented.

and do 

> Can you add a DebugString to the runtime_env_manager_ and print the ref status to worker_pool's debug string?

it in a separate PR?

File: src/ray/common/runtime_env_manager.cc
Comment (rkooo567) [reply src/ray/common/runtime_env_manager.cc:38]: Hmm I see. If it matches to the number of worker processes, that'd be fine

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc:434]: Hmm in this case, I think it is fine!

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc:433]: I think this will help us from committing mistakes, such as somebody add code such as `state.starting_worker_processes.emplace` without adding uri reference. Not sure if the private method is the best choice, but I think this 2 functions should be coupled always. 

Maybe we can just make `OnProcessStarted`, `OnWorkerStarted`, `OnProcessRemoved`?

**Comment by SongGuyang** [comment]
@rkooo567 Where is the `worker lifecycle diagram`? Can you give a link?

**Comment by rkooo567** [comment]
Oh we don’t have one 😂 that’s why I was asking you haha..

**Comment by SongGuyang** [comment]
😂 Let me see. @architkulkarni @edoakes What do you think about the `docstring` mentioned by @rkooo567 ?

**Comment by SongGuyang** [comment]
@architkulkarni @edoakes Can you take a look at this PR again?

File: python/ray/tests/test_runtime_env_working_dir_2.py
Comment (architkulkarni) [reply python/ray/tests/test_runtime_env_working_dir_2.py:308]: Nice, thanks for adding the test with soft_limit=0!  

In the situation I described above, I was actually assuming that soft_limit was large and that the worker was cached.  So step 4 would happen on the same worker that did step 1.  Would it make sense to have a test that runs the above scenario in the case of a larger soft_limit, which is the typical case? The test wouldn't check that GC happens for the task (Like you said, GC won't happen because the worker is cached), the test would just check that step 4 succeeds for the task and that the env wasn't erroneously deleted before step 4. Maybe such a test would be flaky, I don't know exactly how often the soft_limit cap is enforced and when the worker gets disconnected.  In any case, it looks like you covered this situation in your C++ unit test, so I think it's okay.

File: python/ray/tests/test_runtime_env_working_dir.py
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_working_dir.py:68]: This is unfortunate, do you have any thoughts on how we could improve this in the future?

**Review by architkulkarni — COMMENTED** [review]
This looks good to me!  I agree that it would be good to have a docstring somewhere summarizing where we add and delete the URIs.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc:488]: I think I understand, just to make sure could you just clarify the order of these four events?  (1) AddURI in StartWorkerProcess, (2) RemoveURI in MonitorStartingWorkerProcess, (3) AddURI in OnWorkerStarted, (4) RemoveURI in OnWorkerStarted

**Comment by rkooo567** [comment]
cc on https://github.com/ray-project/ray/pull/20789#issuecomment-985303094 I prefer not to merge the PR without this.

**Comment by SongGuyang** [comment]
@rkooo567 Would you like to add it worker pool or runtime env manager?

**Comment by rkooo567** [comment]
To the worker pool!

**Comment by SongGuyang** [comment]
@rkooo567 Thanks for your fast reply. Let me see how to describe.

**Comment by SongGuyang** [comment]
@architkulkarni The python test has been updated. https://github.com/ray-project/ray/pull/20789/commits/694354b0227ad6522ed15324e4671a817f613917

File: python/ray/tests/test_runtime_env_working_dir.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_working_dir.py:68]: Maybe we should use a new job id when reinit?

**Comment by SongGuyang** [comment]
@rkooo567 Do you mean the doc like this? https://github.com/ray-project/ray/pull/20789/commits/9067843d3dbb8de56469f8973f78ee32547c61c3

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:691]: Isn't this having a leak when `num_starting_workers ` > 1? For example, 

`OnWorkerStarted ` will be called per each starting workers, but `RemoveURIReference ` will be called only once when num_starting_workers == 0?

File: src/ray/raylet/worker_pool.h
Comment (rkooo567) [inline src/ray/raylet/worker_pool.h:693]: I think this is close! But I'd like to also see sequence of worker states changes. For example, here are possible events for worker lifecycle (not sure it is correct. It is just an example).

```
OnJobStarted  (+1 reference)
OnJobFinished (-1 reference)
OnProcessStarted   (+1 reference)
OnWorkerTimedOut   (-1 reference)
OnWorkerRegistered (-1 reference -> +1 reference)
DisconnectWorker   (-1 reference)
```

And I'd like to see the state change flow. For example, is it possible to have OnProcessStarted -> DisconnectWorker -> OnWorkerTimedOut? If so, we need to handle the edge case (in this case, the ref count will be -1). 

```
# Only one direction is possible
OnJobStarted -> OnJobFinished
```

```
OnProcessStarted -----------------------------
             |                                | -> DisconnectWorker
OnWorkerTimedOut / OnWorkerRegistered --------
```

If this is the possible state flow, we need to unit test

```
OnProcessStarted -> DisconnectWorker
OnProcessStarted -> OnWorkerTimedOut -> DisconnectWorker
OnProcessStarted -> OnWorkerRegistered -> DisconnectWorker
```

So here, the next question is if the # of calls are guaranteed? Fore example, is it possible?
```
OnProcessStarted -> OnWorkerRegistered X 3 -> DisconnectWorker
```

In a glance, this might be possible when there are more than 1 core worker in a process?

**Review by rkooo567 — COMMENTED** [review]
I will review the newly added part tomorrow!

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:691]: We remove reference in `DisconnectWorker` which is triggered by echo worker instance.

File: src/ray/raylet/worker_pool.h
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:693]: `OnProcessStarted -> DisconnectWorker` I think we shouldn't add this case because `DisconnectWorker ` doesn't happen when there is no worker registration.

Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:693]: `OnProcessStarted -> OnWorkerTimedOut -> DisconnectWorker` I can add this case.

Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:693]: @rkooo567 See this test https://github.com/ray-project/ray/pull/20789/commits/c0178b45f2bb8bb7f047b6a092c5ab52b02baabc#diff-bf9523ff768067f9793ae89d37ff19c53c94ff450e39fd20eaf23a57dd38ccbcR1466

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc]: I have added two functions https://github.com/ray-project/ray/pull/20789/commits/aba42f88018dc314fdde85c303ea1b52c432cbef

**Comment by SongGuyang** [comment]
@rkooo567 Can you see this PR again? We should merge before 1.10.0 cut. I think I already addressed the comments now.

File: src/ray/raylet/worker_pool.h
Comment (rkooo567) [inline src/ray/raylet/worker_pool.h:716]: What does X3 mean here?

File: python/ray/tests/test_runtime_env_conda_and_pip_2.py
Comment (rkooo567) [inline python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Can you add one more python test to test worker registration timeout?

File: python/ray/tests/test_runtime_env_working_dir.py
Comment (rkooo567) [inline python/ray/tests/test_runtime_env_working_dir.py:67]: Can you add TODO?

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc:704]: Sorry what does `echo worker instance.` mean?

File: src/ray/raylet/worker_pool.h
Comment (rkooo567) [inline src/ray/raylet/worker_pool.h:720]: this is nice!

**Review by rkooo567 — COMMENTED** [review]
LGTM. Some comments regarding additional tests. Can you make sure to add them before merging? Also, when you are ready, just ping me. I will approve it and merge it without delay.

File: src/ray/raylet/worker_pool_test.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool_test.cc:1410]: Can you add an unit test for job start / removed?

File: src/ray/raylet/worker_pool_test.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool_test.cc:1411]: Can you separate out first / second part to a different `{}` wrapper 

```
{
  /// first part...
}

{
   /// second part
}
```

or a different test suite?

File: src/ray/raylet/worker_pool_test.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool_test.cc:1410]: Move this ` valid_uris.clear();` to TearDown?

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:704]: Sorry, echo -> each

File: src/ray/raylet/worker_pool.h
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:716]: three workers are started.

Comment (rkooo567) [reply src/ray/raylet/worker_pool.h:716]: So this assumes there are 3 workers per process? Can you mention it somewhere?

File: python/ray/tests/test_runtime_env_conda_and_pip_2.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: I have covered this case in worker pool test. By the way, it is not easy to simulate this in python.

Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: You can simulate this with runtime env I think.

Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Let me see if I can find an example. I am sure there’s a test that simulates thjd

Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: https://github.com/ray-project/ray/blob/ed344341316f29b0c96f9579923a8c106374e04c/python/ray/tests/test_failure_4.py#L440

File: src/ray/raylet/worker_pool_test.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool_test.cc:1410]: Good for me!

File: src/ray/raylet/worker_pool_test.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool_test.cc:1411]: Good for me!

File: python/ray/tests/test_runtime_env_conda_and_pip_2.py
Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: I understand this is covered by cpp test. One concern I have is our worker pool API is not so great that I am not confident it actually conveys the real world workflow (e.g., It is possible some of API is not called on the right order in some scenarios).

**Comment by rkooo567** [comment]
There's the build failure on Windows

File: python/ray/tests/test_runtime_env_conda_and_pip_2.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: So, Do you think I should remove the test from worker pool if we already cover it in python? Actually, I didn't add tests in worker pool before? But @edoakes comment me to add the cpp test https://github.com/ray-project/ray/pull/20789#pullrequestreview-820922744.

Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Why not just adding both?

Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Oh, python worker can't simulate worker process with multiple workers.

Comment (rkooo567) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Oh I meant worker registration failure scenario, not multiple workers

Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: OK, let me see.

**Comment by SongGuyang** [comment]
> There's the build failure on Windows

Thanks！

File: python/ray/tests/test_runtime_env_conda_and_pip_2.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_conda_and_pip_2.py:71]: Add a test here https://github.com/ray-project/ray/pull/20789/commits/884188c1b27d58e412b2f172fd5d7cb7c3e704eb#diff-9a58a116bef057db388a6eef0a46fc55521de1bc9568781a5606792fea7de782R60

File: src/ray/raylet/worker_pool.h
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h]: Add a note here https://github.com/ray-project/ray/pull/20789/commits/884188c1b27d58e412b2f172fd5d7cb7c3e704eb#diff-a88735bc6aa09dd8248a3e73081916b5096a79c4bb7de91f9875d2c5a84ef075R726

File: src/ray/raylet/worker_pool_test.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool_test.cc]: Add test here https://github.com/ray-project/ray/pull/20789/commits/884188c1b27d58e412b2f172fd5d7cb7c3e704eb#diff-bf9523ff768067f9793ae89d37ff19c53c94ff450e39fd20eaf23a57dd38ccbcR1412

**Comment by SongGuyang** [comment]
@rkooo567 Add a issue here https://github.com/ray-project/ray/issues/21108

**Review by rkooo567 — APPROVED** [review]
LGTM

**Comment by rkooo567** [comment]
@architkulkarni please do the final review!

File: python/ray/tests/test_runtime_env_working_dir_3.py
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_working_dir_3.py:50]: What does this do?

**Review by architkulkarni — APPROVED** [review]
Looks great, thanks for adding the tests and the detailed explanation of the refcounting!

**Comment by architkulkarni** [comment]
@SongGuyang Sorry, one last change, it looks like `test_runtime_env_conda_and_pip_2.py` is consistently timing out in CI.  Can you please split it into two test files?  I'm not sure the best way to split it, but I know `test_install_failure_logging` already takes a long time because it repeatedly has to wait 30s for Ray client connection to fail to read the streamed logs (This isn't good, I should find out how to improve this.) So perhaps `test_install_failure_logging` should remain in one file, and the other tests should go in `test_runtime_env_conda_and_pip_3.py`, or maybe you have a different idea.

File: python/ray/tests/test_runtime_env_working_dir_3.py
Comment (SongGuyang) [reply python/ray/tests/test_runtime_env_working_dir_3.py:53]: Used to simulate worker timeout. Add a comment here https://github.com/ray-project/ray/pull/20789/commits/ed4e5f3ecff343f8a71d1bff9a76732efa5e75f7#diff-9a58a116bef057db388a6eef0a46fc55521de1bc9568781a5606792fea7de782R41

**Comment by SongGuyang** [comment]
> @SongGuyang Sorry, one last change, it looks like `test_runtime_env_conda_and_pip_2.py` is consistently timing out in CI. Can you please split it into two test files? I'm not sure the best way to split it, but I know `test_install_failure_logging` already takes a long time because it repeatedly has to wait 30s for Ray client connection to fail to read the streamed logs (This isn't good, I should find out how to improve this.) So perhaps `test_install_failure_logging` should remain in one file, and the other tests should go in `test_runtime_env_conda_and_pip_3.py`, or maybe you have a different idea.

Thank you for this indication. I have separated the test.

**Comment by rkooo567** [comment]
@SongGuyang is it ready to be merged?

**Comment by SongGuyang** [comment]
> @SongGuyang is it ready to be merged?

Yep, we can merge after finish the windows CI.

**Comment by SongGuyang** [comment]
@rkooo567 All checks have passed. Please help to merge.

**Comment by rkooo567** [comment]
This is awesome! Thanks for handling all comments :)!

**Pull request merged**
